RESTserver/src/main/java/com/softserve/if072/restservice/service/ForecastService.java

```diff
@@ -0,0 +1,316 @@
+                while (true) {
+                    if (!(productUsedSpeeds[lastOutlierIndex + 1] < lowerThreeSigmaLimit || productUsedSpeeds[lastOutlierIndex + 1] > upperThreeSigmaLimit)) {
+                        break;
+                    }
+                    otherOutliersCount++;
+                    lastOutlierIndex++;
+                }
+                for (int k = i; k <= lastOutlierIndex; k++) {
+                    productUsedSpeeds[k] = (productUsedSpeeds[i - 1] + productUsedSpeeds[lastOutlierIndex + 1]) / 2;
+                }
+                LOGGER.trace("Find outlier(s) from the index {} to the index {}.", i, lastOutlierIndex);
+                i = lastOutlierIndex + 1;
+            }
+        }
+        LOGGER.trace("Find {} outlier(s) between the beginning and the end of the data set.", otherOutliersCount);
+
+        productStatistics.setProductUsingSpeeds(productUsedSpeeds);
+        productStatistics.setHasOutliers(hasOutliers);
+    }
+
+    /**
+     * Calculate forecasting end date and set it to the ProductStatistics object
+     *
+     * @param productStatistics - object that stores statistical information about product using
+     */
+    private void setEndDate(ProductStatistics productStatistics) {
+        LOGGER.trace("setEndDate method starts to execute.");
+        double mean = productStatistics.getMean();
+        Storage storage = storageDAO.getByProductID(productStatistics.getProductId());
+
+        if (storage == null) {
+            LOGGER.info("There is no \"{}\" with id {} in the user's storage.", productStatistics.getProductName(),
+                    productStatistics.getProductId());
+            productStatistics.setEndDate(null);
+
+            return;
+        }
+        if (storage.getAmount() == 0) {
+            LOGGER.info("There is 0 unit of \"{}\" with id {} in the user's storage.", productStatistics.getProductName(),
+                    productStatistics.getProductId());
+            productStatistics.setEndDate(null);
+
+            return;
+        }
+        productStatistics.setAmountInStorage(storage.getAmount());
+        int amountInStorage = productStatistics.getAmountInStorage();
+        Timestamp endDate = new Timestamp(System.currentTimeMillis() + round(amountInStorage / mean * MILLISECONDS_PER_DAY));
+        productStatistics.setEndDate(endDate);
+        storageDAO.updateEndDate(productStatistics.getProductId(), endDate);
+        LOGGER.info("The end date of the product with id {} has been successfully set in the storage.",
+                productStatistics.getProductId());
+    }
+
+    /**
+     * Calculate forecasting end date of the product in set it to the storage
+     *
+     * @param productId - product unique identifier
+     */
+    public void setEndDate(int productId) {
+        LOGGER.trace("getEndDate method starts to execute.");
+        try {
+            getProductStatistics(productId);
+        } catch (NotEnoughDataException e) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)